File: frontend/src/types/driverIntelligence.ts
export interface DriverTripRead {
  trip_id: string;
  origin: string | null;
  destination: string | null;
  status: string | null;
  pickup_time: string | null;
  actual_delivery_time: string | null;
  delay_minutes: number | null;
  profit_margin: number | null;
}

export interface DriverHoursRead {
  date: string | null;
  trips_count: number | null;
  hours_driven: number | null;
  rest_hours: number | null;
  hos_compliant: boolean | null;
}

export interface DriverBehavior {
  speeding_incidents: number;
  harsh_braking_count: number;
  harsh_accel_count: number;
  violation_count: number;
}

export interface DriverIntelligence {
  driver_id: string;
  driver_name: string | null;
  phone: string | null;
  license_type: string | null;
  license_expiry: string | null;
  experience_years: number | null;
  base_location: string | null;
  rating: number | null;
  status: string | null;
  is_on_trip: boolean;
  license_expiring_soon: boolean | null;
  total_trips: number;
  on_time_rate: number | null;
  avg_delay_minutes: number | null;
  avg_profit_margin: number | null;
  delayed_trips: number;
  behavior: DriverBehavior;
  hos_history: DriverHoursRead[];
  recent_trips: DriverTripRead[];
}

File: backend/app/schemas/driver_intelligence.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict


class DriverTripRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    trip_id: str
    origin: str | None = None
    destination: str | None = None
    status: str | None = None
    pickup_time: datetime | None = None
    actual_delivery_time: datetime | None = None
    delay_minutes: int | None = None
    profit_margin: float | None = None


class DriverHoursRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    date: str | None = None
    trips_count: int | None = None
    hours_driven: float | None = None
    rest_hours: float | None = None
    hos_compliant: bool | None = None


class DriverBehaviorRead(BaseModel):
    speeding_incidents: int = 0
    harsh_braking_count: int = 0
    harsh_accel_count: int = 0
    violation_count: int = 0


class DriverIntelligenceRead(BaseModel):
    driver_id: str
    driver_name: str | None = None
    phone: str | None = None
    license_type: str | None = None
    license_expiry: str | None = None
    experience_years: float | None = None
    base_location: str | None = None
    rating: float | None = None
    status: str | None = None
    is_on_trip: bool
    license_expiring_soon: bool | None = None
    total_trips: int
    on_time_rate: float | None = None
    avg_delay_minutes: float | None = None
    avg_profit_margin: float | None = None
    delayed_trips: int
    behavior: DriverBehaviorRead
    hos_history: list[DriverHoursRead] = []
    recent_trips: list[DriverTripRead] = []
